lib/Dialect/TritonGPU/Transforms/Pipeliner/SoftwarePipeliner.cpp

```diff
@@ -71,9 +78,28 @@ struct PipelinePass : public TritonGPUPipelineBase<PipelinePass> {
     if (this->numStages <= 1)
       return;
     SmallVector<scf::ForOp> loops;
-    getOperation()->walk([&](scf::ForOp forOp) { loops.push_back(forOp); });
+
+    // Process the scf.for operations from the inside-out.
+    getOperation()->walk<WalkOrder::PostOrder>([&](scf::ForOp forOp) {
+      loops.push_back(forOp);
+    });
+
     for (scf::ForOp forOp : loops) {
-      pipelineLoop(forOp, numStages);
+      // Parents parents of already-pipelined for loops are null.
+      if (!forOp)
+        continue;
+
+      FailureOr<scf::ForOp> maybeForOp = pipelineLoop(forOp, numStages);
+
+      // If this was pipelined don't pipeline any of the parents.
+      if (succeeded(maybeForOp)) {
+        forOp = *maybeForOp;
+        while ((forOp = forOp.getOperation()->getParentOfType<scf::ForOp>())) {
+          if (scf::ForOp* it = std::find(loops.begin(), loops.end(), forOp)) {
+            *it = nullptr;
```

Comment:
again this makes assumptions on the order in which forop are processed. I think we should just skip any loop than contains nested loops. We already scan through ops line 40. Can we extend to include this pre-condition.